[Ramanathan Kumarappan] iP

Ramanathan0908 opened the pull request:
# Duke
> My mind wants to have ideas, but does not hold them
Duke is simple and good
- text-based
- can be picked up fast
- ~~FAST~~ **_ULTRA FAST_**

All you need to do is,
1. download it from [here](https://github.com/Ramanathan0908/ip/releases/download/A-Jar/duke.jar)
2. launch it
3. use it! :shipit: 

Features
- [x] keep track of todos
- [x] keep track of deadlines and events
- [ ] have an gui

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public static void main(String[] args) {
        UI ui = new UI();
        try {
            TaskList taskList = new TaskList();
            Storage storage = new Storage(taskList);
            Parser parser = new Parser();
            
            String commandResult;
            ui.showIntroMsg();
            do {
                String userInput = ui.getUserCommand();
                commandResult = parser.parseInput(userInput, taskList);
                storage.saveTaskList(taskList);
                ui.showToUser(commandResult);
            } while (!commandResult.equals("exit"));
            ui.showOutroMsg();
        } catch (DukeException e) {
            ui.showToUser(e.getMessage());
        }
    }
```

Commit:
Add Gradle support

Ramanathan0908 committed:
implemented duke level 1

Ramanathan0908 committed:
implemented level 2 of Duke

Ramanathan0908 committed:
implemented Level 3 of Duke

Ramanathan0908 committed:
refactored out the task classes to separate files

Ramanathan0908 committed:
implemented support for tracking three types of tasks

Ramanathan0908 committed:
Automated text UI testing

Ramanathan0908 committed:
implemented error handling

Ramanathan0908 committed:
added support for deleting tasks

Ramanathan0908 committed:
Add ability to save and load data

Ramanathan0908 committed:
Add proper time support for deadline and event tasks

Ramanathan0908 committed:
Merge branch 'branch-Level-7'
merge level-7 branch into master

Ramanathan0908 committed:
Merge branch 'branch-Level-8'
merge level-8 branch into master

Ramanathan0908 committed:
Use more OOP

Ramanathan0908 committed:
Refactor classes into appropriate packages

Ramanathan0908 committed:
Merge branch 'add-gradle-support' of github.com:nus-cs2103-AY2223S1/ip
Add gradle support.

Ramanathan0908 committed:
Use gradle.

Ramanathan0908 committed:
Add Junit Tests

Ramanathan0908 committed:
Add JavaDoc comments.

Ramanathan0908 committed:
Follow the coding standard

Ramanathan0908 committed:
Give users a way to find a task

Ramanathan0908 committed:
resolved merge conflict between master and branch-A-CodingStandard

Ramanathan0908 committed:
Merge branch 'branch-Level-9'
merge branch-Level-9 into master

craeyeons commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,42 @@
+package duke;
+
+import duke.dukeExceptions.DukeException;
+import org.junit.jupiter.api.BeforeEach;
```
I think i'd like to see a line between the two import statements, so I can clearly see a line between duke imports and junit imports.

craeyeons commented on src/test/java/duke/tasks/DeadlineTest.java:
```diff
@@ -0,0 +1,24 @@
+package duke.tasks;
+
+import duke.dukeExceptions.DukeException;
+import org.junit.jupiter.api.BeforeEach;
```
I think i'd like to see a line between the two import statements, so I can clearly see a line between duke imports and junit imports.

craeyeons commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    /**
+     * Retrieves the specified Task from TaskList.
+     * @param i - An int specifying the index of the Task in the TaskList.
+     * @return The Task at the given index.
+     */
```
I don't think I like that you added a dash ("-") before writing the description for each parameter. I have seen this in other places as well.

To add, I realised some inconsistency in adding dashes in return statements, e.g. in this case you didn't, but you did at getUserCommand() of Ui.java.

craeyeons commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,128 @@
+     * @return - The resultant String of executing the command.
+     * @throws DukeException - When there is an error with the command.
+     */
+    public String parseInput(String input, TaskList tl) throws DukeException {
```
I don't think I like tl as the variable name of a TaskList instance. I understand that you have derived it from the abbreviation of TaskList, however, considering the length of the code, tl may be a little too succinct.

craeyeons commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    /**
+     * Lists the Tasks in the TaskList.
+     * @return A String representing the Tasks in the TaskList.
```
I'd like a line between the method description and all of the other tags. I've seen it in other places as well, maybe you can considering changing.

craeyeons commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,82 @@
+     * @param by - The date and time for the deadline (in YYYY-MM-DD HHMM format).
+     * @throws DukeException - When the description or by date is invalid.
+     */
+    public Deadline(String description, String by) throws DukeException {
```
I like how you let time be an optional parameter when setting a deadline.

kelvinou01 replied:
Now, this constructor has two responsibilities: parsing the `by` string and creating the `Deadline`. Would defining another method for the parse, and making the constructor take a `LocalDate` as a parameter make things clearer?

craeyeons reviewed (commented):
Good job overall! Just a few nits to fix!

kelvinou01 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,31 @@
+                commandResult = parser.parseInput(userInput, taskList);
+                storage.saveTaskList(taskList);
+                ui.showToUser(commandResult);
+            } while (!commandResult.equals("exit"));
```
I'm not sure if I like the conditional statement. `"exit"` is a magic string that may be broken easily by a future programmer who isn't as careful as you.

kelvinou01 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,128 @@
+        String arguments = tokens.length == 2 ? tokens[1] : null;
+        
+        switch (command) {
+            case "bye": 
```
Remember, `case` shouldn't be indented!

kelvinou01 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,128 @@
+                if (deadlineArgs.length == 1) {
+                    throw new DukeException("Deadline requires a /by date");
+                }
+                System.out.println(deadlineArgs[1]);
```
I think I would use the UI for displaying to the User. Using print would break the abstraction barrier.

kelvinou01 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,128 @@
+                return "I've added: " + newTask + " you have " + tl.getTaskListSize() +" tasks left";
+            }
+            case "mark": {
+                try {
```
I'm not sure if I like how much code is wrapped in the `try` block. I could be wrong, but I think its possible to minimise further the number of lines wrapped in `try`.

kelvinou01 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,128 @@
+                }
+            }
+            case "find": {
+//                try {
```
Oops, I think you forgot to remove these comments before pushing!

kelvinou01 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,83 @@
+                while (s.hasNext()) {
+                    String task = s.nextLine();
+                    switch (task.charAt(0)) {
+                        case 'T':
```
I'd like to see less nesting here. Perhaps we can break up the method into two or three methods, each with a single responsibility? 

(e.g. I would write a method to handle mapping `String task` to the correct `addTask` call)

kelvinou01 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,83 @@
+     * @throws DukeException - When there is an error saving storage.
+     */
+    public void saveTaskList(TaskList tl) throws DukeException {
+        try {
```
Similarly, I think it would be possible to reduce the number of lines wrapped by the `try` block.

kelvinou01 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,66 @@
+    /**
+     * Constructor for TaskList.
+     */
+    TaskList() {
```
Should this be `public`?

kelvinou01 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,66 @@
+     * Lists the Tasks in the TaskList.
+     * @return A String representing the Tasks in the TaskList.
+     */
+    public String listTasks() {
```
This method is returning the string representation of your tasks. Is there another name for this method that better captures what its doing?

kelvinou01 commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,80 @@
+ * @author Ramanathan Kumarappan
+ */
+public class UI {
+    private static final String LS = System.lineSeparator();
```
Would a more descriptive name help?

kelvinou01 commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,80 @@
+ * 
+ * @author Ramanathan Kumarappan
+ */
+public class UI {
```
Just a minor nitpick, but perhaps `Ui` would better conform to CamelCase?

kelvinou01 commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,80 @@
+        this.OUT = System.out;
+    }
+    
+    private boolean shouldIgnore(String rawInput) {
```
Great implementation, but I find it difficult to discern what this method does by the name. 

I would name it based on what it is directly doing (taking a string and seeing if its actually just white space).

kelvinou01 commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,80 @@
+    /**
+     * Constructor for UI.
+     */
+    private final Scanner IN;
+    private final PrintStream OUT;
+    
+    UI() {
+        this.IN = new Scanner(System.in);
+        this.OUT = System.out;
+    }
```
Interesting decision to encapsulate the `Scanner` and `PrintStream` into variables. Would you share your thought process?

Ramanathan0908 replied:
Hey, sorry for such a late response! but I wanted the flexibility in the future to change the input and output sources without too much hassle, hence I encapsulated them into variables

kelvinou01 commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,82 @@
+     * @return A deadline Task.
+     * @throws DukeException - When Deadline cannot be reconstructed from the given string.
+     */
+    public static Deadline taskFromSave(String saveString) throws DukeException {
```
I'm not sure if I like this method name.

kelvinou01 commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,81 @@
+     * @return The event in a save string format.
+     */
+    @Override
+    public String saveString() {
```
Again, I think this method name can be made clearer.

kelvinou01 commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,57 @@
+     * 
+     * @param description - The description of the Task.
+     */
+    Task(String description) {
```
I'd be explicit about the accessibility of this constructor.

Ramanathan0908 committed:
Add GUI for Duke

fongyj commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,47 @@
+import javafx.scene.control.Label;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
+
```
I think you might have forgotten to include javadocs for this class and some of the methods in this class too

fongyj commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,47 @@
+    private Label text;
+    private ImageView displayPicture;
+
+    public DialogBox(Label l, ImageView iv) {
```
I think the argument names "l" and "iv" should be more descriptive

fongyj commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,9 @@
+
+import javafx.application.Application;
+
+public class Launcher {
```
Similarly this class seems to be missing javadocs

fongyj reviewed (commented):
I looked through the comments left by the other reviewers and I feel they were already quite thorough with the review. Apart from the issues raised by them, I think there are also some issues with the scripts that were pushed more recently. Other than that, the coding style looks good 👍

Ramanathan0908 committed:
added assertions to document important assumptions

Ramanathan0908 committed:
Code quality improvements

Changes were made to the javadocs for methods, hypens were removed between
parameters and their descriptions. A line break was added between method
description and tags.

Switch statements were properly indented according to the style guide.

Ramanathan0908 committed:
Implement sorting by chronological order of tasks.

Add new cmd "sort" in parser.

Implemented comparable interface for Tasks that compares tasks
by date and time if available (Tasks with no date or time gets lowest
priority).

Ramanathan0908 committed:
Merge pull request #1 from Ramanathan0908/branch-A-Assertions

added assertions to document important assumptions

Ramanathan0908 committed:
Merge branch 'master' of github.com:Ramanathan0908/ip into branch-C-Sort

Ramanathan0908 committed:
Merge pull request #2 from Ramanathan0908/branch-A-CodeQuality

Code quality improvements

Ramanathan0908 committed:
resolved merge conflicts

Ramanathan0908 committed:
Merge pull request #3 from Ramanathan0908/branch-C-Sort

Branch c sort

Ramanathan0908 committed:
Add User guide

Ramanathan0908 committed:
Add Ui image

Ramanathan0908 committed:
Update credits for reused code
